index.js

```diff
@@ -54,25 +143,67 @@ MediaSourceStream.prototype._write = function (chunk, encoding, cb) {
     return cb(new Error('Cannot append buffer while source buffer updating'))
   }
 
-  self._sourceBuffer.appendBuffer(chunk)
+  try {
+    self._sourceBuffer.appendBuffer(toArrayBuffer(chunk))
+  } catch (err) {
+    self.emit('error', err)
+    return
+  }
   self._cb = cb
 }
 
 MediaSourceStream.prototype._flow = function () {
   var self = this
+
+  if (self.destroyed || !self._sourceBuffer || self._sourceBuffer.updating) {
+    return
+  }
+
+  if (self._mediaSource.readyState === 'open') {
+    // check buffer size
+    if (self._getBufferDuration() > self._bufferDuration) {
+      return
+    }
+  }
+
   if (self._cb) {
-    self._cb(null)
+    var cb = self._cb
+    self._cb = null
+    cb()
   }
 }
 
-function getType (extname) {
-  if (!extname) return null
-  if (extname[0] !== '.') extname = '.' + extname
-  return {
-    '.m4a': 'audio/mp4; codecs="mp4a.40.5"',
-    '.m4v': 'video/mp4; codecs="avc1.640029, mp4a.40.5"',
-    '.mp3': 'audio/mpeg',
-    '.mp4': 'video/mp4; codecs="avc1.640029, mp4a.40.5"',
-    '.webm': 'video/webm; codecs="vorbis, vp8"'
```

Comment:
Glad you got rid of this hackiness. This can just go into `render-media` for the fallback-of-last-resort case.